oauthenticator/gitlab.py

```diff
@@ -168,36 +168,38 @@ class GitLabOAuthenticator(OAuthenticator):
             return None
 
 
-    async def _check_group_whitelist(self, user_id, access_token):
+    async def _check_group_whitelist(self, user_id, username, access_token):
         http_client = AsyncHTTPClient()
         headers = _api_headers(access_token)
         # Check if user is a member of any group in the whitelist
         for group in map(url_escape, self.gitlab_group_whitelist):
-            url = "%s/groups/%s/members/%d" % (GITLAB_API, group, user_id)
+            url = "%s/groups/%s/members/all?query=%s" % (GITLAB_API, group, username)
             req = HTTPRequest(url, method="GET", headers=headers)
             resp = await http_client.fetch(req, raise_error=False)
-            if resp.code == 200:
-                return True  # user _is_ in group
+            if resp.code == 200 and resp.body:
+                resp_json = json.loads(resp.body.decode('utf8', 'replace'))
+                for user in resp_json:
+                    if user['id'] == user_id and user['username'] == username:
+                        return True  # user _is_ in group
         return False
 
 
-    async def _check_project_id_whitelist(self, user_id, access_token):
+    async def _check_project_id_whitelist(self, user_id, username, access_token):
         http_client = AsyncHTTPClient()
         headers = _api_headers(access_token)
         # Check if user has developer access to any project in the whitelist
         for project in self.gitlab_project_id_whitelist:
-            url = "%s/projects/%s/members/%d" % (GITLAB_API, project, user_id)
+            url = "%s/projects/%s/members/all?query=%s" % (GITLAB_API, project, username)
             req = HTTPRequest(url, method="GET", headers=headers)
             resp = await http_client.fetch(req, raise_error=False)
 
-            if resp.body:
+            if resp.code == 200 and resp.body:
                 resp_json = json.loads(resp.body.decode('utf8', 'replace'))
-                access_level = resp_json.get('access_level', 0)
-
-                # We only allow access level Developer and above
-                # Reference: https://docs.gitlab.com/ee/api/members.html
-                if resp.code == 200 and access_level >= 30:
-                    return True
+                for user in resp_json:
+                    if user['id'] == user_id and user['username'] == username:
+                        # We only allow access level Developer and above
+                        # Reference: https://docs.gitlab.com/ee/api/members.html
+                        return user['access_level'] >= 30
```

Comment:
Same question as above regarding the redundancy, and :heart: for including a good reference pointer to the access_level details!

Reply:
To be fair, the link to the docs was already there 😉